src/cursor/abstract_cursor.ts

```diff
@@ -1032,41 +1037,57 @@ export abstract class AbstractCursor<
         return this.timeoutContext?.refreshed();
       }
     };
-    try {
-      if (
-        !this.isKilled &&
-        this.cursorId &&
-        !this.cursorId.isZero() &&
-        this.cursorNamespace &&
-        this.selectedServer &&
-        !this.cursorSession.hasEnded
-      ) {
-        this.isKilled = true;
-        const cursorId = this.cursorId;
-        this.cursorId = Long.ZERO;
-
-        await executeOperation(
-          this.cursorClient,
-          new KillCursorsOperation(cursorId, this.cursorNamespace, this.selectedServer, {
-            session: this.cursorSession
-          }),
-          timeoutContextForKillCursors()
-        );
+
+    const withEmitClose = async (fn: () => Promise<void>) => {
+      try {
+        await fn();
+      } finally {
+        this.emitClose();
       }
-    } catch (error) {
-      squashError(error);
-    } finally {
+    };
+
+    const close = async () => {
+      // if no session has been defined on the cursor, the cursor was never initialized
+      // or the cursor was re-wound and never re-iterated.  In either case, we
+      //   1. do not need to end the session (there is no session after all)
+      //   2. do not need to kill the cursor server-side
+      const session = this.cursorSession;
+      if (!session) return;
+
       try {
-        if (this.cursorSession?.owner === this) {
-          await this.cursorSession.endSession({ error });
-        }
-        if (!this.cursorSession?.inTransaction()) {
-          maybeClearPinnedConnection(this.cursorSession, { error });
+        if (
+          !this.isKilled &&
+          this.cursorId &&
+          !this.cursorId.isZero() &&
+          this.cursorNamespace &&
+          this.selectedServer &&
+          !session.hasEnded
+        ) {
+          this.isKilled = true;
+          const cursorId = this.cursorId;
+          this.cursorId = Long.ZERO;
+
+          await executeOperation(
+            this.cursorClient,
+            new KillCursorsOperation(cursorId, this.cursorNamespace, this.selectedServer, {
+              session
+            }),
+            timeoutContextForKillCursors()
+          );
         }
+      } catch (error) {
+        squashError(error);
       } finally {
-        this.emitClose();
+        if (session.owner === this) {
+          await session.endSession({ error });
+        }
+        if (!session?.inTransaction()) {
```

Comment:
can session become cleared in the course of this function? (asking since we have an early return on null in L1055)
for consistency, either 1081 needs a `?` or this line does not need it

Reply:
Nope, `session` is `const`.  Fixed